stdenv-generic: add meta attribute checking

copumpkin opened the pull request:
This is turned off by default but I think we should fix all packages to respect it and then turn it on by default. To turn it on, just put `checkMeta = true;` into your nixpkgs config.

cc @shlevy @dezgeg @edolstra

###### Motivation for this change

I got sick of constantly seeing typos and accidental type errors in our meta tags, and people never notice them until Hydra complains. I'd probably eventually start adding a few more checks (like that `platforms` members are in an enum and so on) but this should get us 95% of the value.

It's turned off by default because there are tons of violations, which I'd propose fixing incrementally in separate commits. The very first failure is in binutils, which has a `priority` field as a string 😄 

###### Things done

Nothing. I just tried evaluating `stdenv` and got some failures, so there's work to be done 😁

N.B: this is _not_ a mass rebuild, and only affects evaluation

copumpkin committed:
stdenv-generic: add meta attribute checking

This is turned off by default but I think we should fix all packages to
respect it and then turn it on by default

shlevy reviewed (approved):
:ship: :ship: :ship:

copumpkin commented:
Okay, now that this is merged, if anyone wants to join me in making small commits to fix meta attributes, please feel free 😄

copumpkin commented:
I ran `hydra-eval-jobs` per @cleverca22's suggestion and I think I've fixed all remaining issues, so I switched `checkMeta` to `true` by default. Please let me know or switch to `false` if you object, but I'd rather sneak it in before others break more stuff.

copumpkin commented:
The two commits were https://github.com/NixOS/nixpkgs/commit/1a4ca220e15480818b78041d46bf0ca438671d33 and https://github.com/NixOS/nixpkgs/commit/90b9719f4fc95e54400a66bffcc8044c568cfa4b

edolstra commented:
The thing is that `meta` is intended as an arbitrary name/value map. So it's perfectly fine to include attributes that are not "known".

copumpkin commented:
@edolstra my sense is that the reason you'd add stuff to meta rather than a comment is that you want it to be structured for some form of machine consumption. If you want a machine to consume it, you generally want it to have a reasonably well defined schema. So it doesn't seem too burdensome to ask people wanting to add a new meta attribute to add it to the central schema first (and it'll be easy for them to tell if they miss it because it'll fail eagerly now)

On the evaluation speed, I don't know and haven't measured yet. Where can I find evaluation times in Hydra? I guess if it slows it down excessively we can turn it off, but given the literally dozens of edits I made to `nixpkgs` here to fix stupid typos I found with this feature, this is solving a real problem, so I'm loath to privilege speed of Hydra evaluation (which is rare and not something users -facing, and comparatively small compared to build time) over the system silently ignoring typo'd meta tags.

7c6f434c commented:
@copumpkin speed of evaluation is something visible via `nix-env -qa`; and also `nix-env -i` which shouldn't be used ever but oh well…

If you want a quick measurement, try measuring `test-eval-release.sh`, I think it should be a close proxy.

By the way, maybe the check should be done inside the tarball build, like the warning checks? Tarball failures block the channel, so they do get noticed…

(I did have to do some fixes because of this check being introduced, and I do like this check in general)

copumpkin commented:
We could put it in the tarball I guess but feedback loop time makes a
difference. People still won't ever notice if they wrote maitnainers
instead of maintainers (or home/website instead of homepage), as someone
else will probably notice the tarball is blocked and fix it. Of course we
can use more human ways of spreading the message around like adding
instructions for turning this on in the PR template but I'd rather default
to a mode that doesn't silently ignore silly mistakes if we can.

Anyway, I'm not near my computer now but will try to measure eval time
later, if Hydra isn't already doing it somewhere.
On Mon, May 1, 2017 at 11:47 Michael Raskin <notifications@github.com>
wrote:

> @copumpkin <https://github.com/copumpkin> speed of evaluation is
> something visible via nix-env -qa; and also nix-env -i which shouldn't be
> used ever but oh well…
>
> If you want a quick measurement, try measuring test-eval-release.sh, I
> think it should be a close proxy.
>
> By the way, maybe the check should be done inside the tarball build, like
> the warning checks? Tarball failures block the channel, so they do get
> noticed…
>
> (I did have to do some fixes because of this check being introduced, and I
> do like this check in general)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/25304#issuecomment-298358688>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKP7_mIqhHowMeiigYamcK8ZADkyfJks5r1f6MgaJpZM4NMBSF>
> .
>

7c6f434c commented:
>can use more human ways of spreading the message around like adding
>instructions for turning this on in the PR template but I'd rather default

Also tagging people who broke tarball eval with an explanation of the
problem seems to create some progress.

Maybe a special environment variable that makes stdenv return strict or
lax mkDerivation could be a good solution? If we set the lax variable 
value for Hydra and `nix-env -qa`, but not `test-eval-release.sh`,
tarball build or `nix-build`/`nix-env -iA`, we could have the best of
both worlds…

edolstra commented:
http://hydra.nixos.org/job/nixpkgs/trunk/metrics#tabs-charts

Note the jump for e.g. `nix-env.qa.values` around April 29 from around 2 million to 3.5 million. Or `nix-env.qa.maxresident` which went from 654 MB to 732 MB.

edolstra commented:
@7c6f434c Yeah, doing this kind of lint check in the Nixpkgs tarball job sounds good.

copumpkin commented:
Yeah, so I guess it did go up a bit. The real question to me is what we're optimizing for. I'd obviously prefer what I'd call "developer efficiency", but I'm not the one paying for or maintaining Hydra hardware. At the same time, losing a few seconds (or <100MB of RAM) with each full-nixpkgs hydra evaluation when builds take hours or even days still seems pretty small.

I do realize that the evaluation time matters every time we do `nix-env` or `nix-shell` for users too, but even there it feels like it's small compared to downloading packages from the binary cache and tiny compared to building them.

Anyway, I can't really make the call, but if I were running the project, I'd probably leave this turned on by default and accept slightly slower Hydra eval times.

7c6f434c commented:
@copumpkin @edolstra and what about environment-variable-controlled middle ground?

copumpkin commented:
It's already controlled by a config flag. The question is more whether we want it on by default or not.

7c6f434c commented:
My point is that `nix-build` and `nix-env -qa` with the default config should behave differently.

On Hydra the check should be just switched off in evaluation and moved to the tarball build.

copumpkin commented:
Ah, so we'd have `nix-build` always check meta, `nix-env` wouldn't check it, and Hydra would only check when building the tarball? I guess that isn't awful, although it does (by design) make things behave a little differently in potentially confusing ways. How do I distinguish between `nix-build` and `nix-env` (also, what would `nix-shell` do?) with an env var?

7c6f434c commented:
I would set an environment variable; and `nix-shell` is closer to `nix-build`; I actually would make `nix-env -i` invocations of the `nix-build` kind and only `-qa` lax.

There would be some slight difference in behaviour, true, but at least in case of success the hashes would be the same, so not that bad… I doubt we can go more confusing that the current `broken`/`override` interaction.

copumpkin commented:
Oh, you mean developers would need to set an environment variable when developing? That feels like defaulting it to off, which I didn't want.

I was hoping for something a bit like `IN_NIX_SHELL` that would let me distinguish between `nix-build` and `nix-env`.

My issue is just that the people who are typoing all their meta tags (often in consistently bad ways, like `website`, `home`, etc.) today probably aren't following our PRs that closely and won't know that I've added this feature. So the people who need it the most won't know that they need it.

copumpkin commented:
I'll just turn it off by default for now, since I think I missed some of the python packages which don't get evaluated by Hydra. I still want to turn it on by default, but can get more community input or make an RFC or something like that if you think this is too impactful.

@edolstra where are you leaning? Did my "unstructured random data in `meta` is useless" point make sense? Do you think the performance issues are actually a big deal?

Trying to gauge how controversial this is, basically, to decide whether an RFC is worthwhile.

copumpkin commented:
I switched the default to `false` again (for now, I hope) in https://github.com/NixOS/nixpkgs/commit/f3a05a0fb36bc63ce00e47859b0fea4914eb482c

7c6f434c commented:
No-no, I meant that `nix*` tools would set the variable that distinguishes mass operations from narrowly targeted ones. I would hope that adding a new package includes `nix-build`, which is a targeted operation and would do the strict check.

7c6f434c commented:
Distinguishing mass operations and targeted operations would be also useful for deprecation warnings. 

For example, @oxij wonders is there a way to display notices that there are two different `torbrowser` packages (from source and binary) now.

oxij commented:
> For example, @oxij wonders is there a way to display notices that there are two different `torbrowser` packages (from source and binary) now.

While my wishes are related, its not exactly the same use case, AFAICS.

E.g. I don't want users of `tor-browser-bundle-bin` to see the warning,
only those who use the old `torbrowser` attr.

On a second thought, though, its probably possible to do something like

~~~~
torbrowser = withWarning "..." tor-browser-bundle-bin;
~~~~

in `all-packages.nix` with `withWarning` generating new derivation with
a warning in meta. Then I would even vote for evaluating such warning in
`nix-env -i`.

On a related note. Having a way to catch `assert` failures in
derivations would be nice too.

I want to fuzz boolean package options and check that packages build
with all possible combinations of them, directly in nix, but nix has no
`catch`.

Rewriting assertions to be a list in `meta` that gets inspected only on
derivation build attempt would work for that use case too.

8573 commented on pkgs/stdenv/generic/default.nix:
```diff
@@ -202,6 +205,32 @@ let
+        branch = str;
+        homepage = str;
+        downloadPage = str;
+        license = either (listOf lib.types.attrs) lib.types.attrs;
```
[The current documentation][1] says that using strings for `license` is "valid", though "frowned upon". I suppose the manual should be updated?

[1]: <https://github.com/NixOS/nixpkgs/blob/7f7ebe5bf6ecb6fcb3c5d603c8f2e5b927ad0750/doc/meta.xml#L134-L169>

8573 commented on pkgs/stdenv/generic/default.nix:
```diff
@@ -202,6 +205,32 @@ let
+
+        # Weirder stuff that doesn't appear in the documentation?
+        version = str;
+        updateWalker = bool;
```
`updateWalker` seems to be [documented][1].

[1]: <https://github.com/NixOS/nixpkgs/blob/7f7ebe5bf6ecb6fcb3c5d603c8f2e5b927ad0750/doc/meta.xml#L238-L246>
